---
title: "Cassie Winn R Assignment"
output: html_document
---

# Part I
## Data Inspection
### Read In Files

```{r }
library(tidyverse)
genotypes <- read_tsv("./data/fang_et_al_genotypes.txt")
snp_position <- read_tsv("./data/snp_position.txt")
```
### Determine File Size
Use file.info to get size of file in bytes, whether the file is a directory, last status change and last access times, user ID of the file's owner, username, group name, and other info.
```{r}
file.info("./data/fang_et_al_genotypes.txt")
file.info("./data/snp_position.txt")
```

If you just wanted file size in bytes use the function file.size()
```{r}
file.size("./data/fang_et_al_genotypex.txt")
file.size("./data/snp_position.txt")
```


### Determine Number of Lines 
Simply type the name of the dataframe and it will give the dimensions for lines x columns (example: "A tibble: 983 x 15")

```{r}
genotypes
snp_position
```
Or, use the function nrow() to give the number of rows/lines in a dataframe
```{r}
nrow(genotypes)
nrow(snp_position)
```

Or, use the function dim() to get rows and columns
```{r}
dim(genotypes)
dim(snp_position)
```


### Determine Number of Columns
Use ncol() function to determine number of columns in a dataframe
```{r}
ncol(genotypes)
ncol(snp_position)
```


## Data Processing
### Extracting Info from Files

First, I extracted the maize and teosinte data and made two new corresponding files

```{r}
maize_genotypes <- subset(genotypes, Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR")
teosinte_genotypes <- subset (genotypes, Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
```

### Transpose 
To transpose the data we can use function t()
```{r}
transposed_maize <- as.data.frame(t(maize_genotypes))
transposed_teosinte <- as.data.frame(t(teosinte_genotypes))
```


### Format Genotype Files and Sort
The transposed data does not include the header in the first column, so we need to add that in. 

```{r}
if(!require("tibble")) install.packages("tibble")
library(tibble)

names(transposed_maize) <- lapply(transposed_maize[1,], as.character)
transposed_maize <- transposed_maize [-1,]

names(transposed_teosinte) <- lapply(transposed_teosinte[1, ], as.character)
transposed_teosinte <- transposed_teosinte[-1,]

transposed_maize <- rownames_to_column(transposed_maize, var="SNP_ID")
transposed_teosinte <- rownames_to_column(transposed_teosinte, var="SNP_ID")

```

We then sort the files by SNP_ID. To do that use the dplyr function arrange()
```{r}
if(!require("dplyr")) install.packages("dplyr")
library(dplyr)

transposed_teosinte <- arrange(transposed_teosinte, SNP_ID) 
transposed_maize <- arrange(transposed_maize, SNP_ID)
```

### Format and Sort SNP File
First, I cut out the columns of snp_position.txt that are needed (SNP_ID, Chromosome and Position) and compiled them in one file in the order desired.

```{r}
snp_ID_chro_pos <- snp_position[, c("SNP_ID", "Chromosome", "Position")]
```

Before the SNP file can be joined with genotype files, it must be sorted by SNP_ID as well:

```{r}
snp_ID_chro_pos <- arrange(snp_ID_chro_pos, SNP_ID) 
```

### Join SNP and Genotype Files
Join the SNP and Genotype files using merge()
```{r}
teosinte_join <- merge(snp_ID_chro_pos,transposed_teosinte, by.x="SNP_ID", by.y="SNP_ID", all = TRUE)

maize_join <- merge(snp_ID_chro_pos,transposed_maize, by.x="SNP_ID", by.y="SNP_ID", all = TRUE)

```

We then need to remove the two rows that came from the genotype files that are not informative (Group and JG_OTU). 
```{r}
teosinte_join <- subset(teosinte_join, !(teosinte_join$Chromosome == "unknown") | !(teosinte_join$Chromosome =="multiple"))

maize_join <- subset(maize_join, !(maize_join$Chromosome == "unknown") | !(maize_join$Chromosome == "multiple"))
```

### Extract Data for Input Files
Create 10 files (1 for each chromosome) where SNPs are ordered based on increasing position and missing data encoded by this symbol: ? 

First, arrange the files with position sorted in ascending order
```{r}
teosinte_join_ascend <- arrange(teosinte_join, Position)
maize_join_ascend <- arrange(maize_join, Position)
```

To create the 10 files (1 for each chromosome) we use a for loop and write.csv() to create the files.
```{r}
setwd("./files/Maize")
for (i in 1:10) {
  x <- maize_join_ascend[maize_join_ascend$Chromosome == i,]
#we can create a new CSV file for each chromosome
  write.csv(x, sprintf("maize_increase_chrom_%d.csv", i), row.names = F)
}

```

Do the same for teosinte files.
```{r}
setwd("./files/Teosinte")
for (i in 1:10) {
  x <- teosinte_join_ascend[teosinte_join_ascend$Chromosome == i,]
#we can create a new CSV file for each chromosome
  write.csv(x, sprintf("teosinte_increase_chrom_%d.csv", i), row.names = F)
}
```

Create 10 files (1 for each chromosome) where SNPs are ordered based on decreasing position and missing data is encoded by this symbol: - 

First, I order the dataframe in decreasing order
```{r}
teosinte_join_descend <- arrange(teosinte_join, desc(Position))
maize_join_descend <- arrange(maize_join, desc(Position))
```

Next, I change the symbol ? to - and again create the files for each chromosome.
```{r}
teosinte_join_descend[] <- lapply(teosinte_join_descend, as.character)
teosinte_join_descend[teosinte_join_descend == '?/?'] <- '-/-'

maize_join_descend[] <- lapply(maize_join_descend, as.character)
maize_join_descend[maize_join_descend == '?/?'] <- '-/-'

```

```{r}
setwd("./files/Maize")
for (i in 1:10) {
  x <- maize_join_descend[maize_join_descend$Chromosome == i,]
#we can create a new CSV file for each chromosome
  write.csv(x, sprintf("maize_decrease_chrom_%d.csv", i), row.names = F)
}
```

```{r}
setwd("./files/Teosinte")
for (i in 1:10) {
  x <- teosinte_join_descend[teosinte_join_descend$Chromosome == i,]
#we can create a new CSV file for each chromosome
  write.csv(x, sprintf("teosinte_decrease_chrom_%d.csv", i), row.names = F)
}
```


# Part 2

Load required packages
```{r}
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)

if(!require("tibble")) install.packages("tibble")
library(tibble)

if (!require("reshape2")) install.packages("reshape2")
library(reshape2)

if(!require("dplyr")) install.packages("dplyr")
library(dplyr)

if (!require("plyr")) install.packages("plyr")
library(plyr)

```

First, I transpose the genotypes file, add in the header and join it with the snp_position file to have a dataset ready for graphing.

```{r}
genotypes_transpose <- as.data.frame(t(genotypes))

names(genotypes_transpose) <- lapply(genotypes_transpose[1, ], as.character)
genotypes_transpose <- genotypes_transpose[-1,]
genotypes_transpose <- rownames_to_column(genotypes_transpose, var="SNP_ID")

genotypes_trans_join <- merge(snp_position,genotypes_transpose, by.x="SNP_ID", by.y="SNP_ID", all = TRUE)
```
### SNPs per chromosome
Plot the total number of SNPs in our dataset on each chromosome. What groups contribute most of these SNPs?

First, create a vector with the chromosomes in order.
```{r}
genotypes_trans_join$Chromosome <- factor(genotypes_trans_join$Chromosome, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown", "NA"))
```

Create the graph using ggplot()
```{r}
ggplot(genotypes_trans_join) + geom_bar(aes(genotypes_trans_join$Chromosome)) + xlab("Chromosome") + ylab("Total Number of SNPs")
```

To look at which groups contribute most of these SNPs, use ggplot() to look at the original genotypes from fang_et_al_genotypes.txt
```{r}
ggplot(genotypes) + geom_bar(aes(genotypes$Group)) + xlab("Group") + ylab("Total Number of SNPs")

```

### Missing Data and Amount of Heterozygosity

Using melt() we put the individual value of each SNP in each of the sample_id and JG_OTU
```{r}
header <- colnames(genotypes)[-c(1:3)]
genotypes_melt <- melt(genotypes, measure.vars = header)
```

Change the symbol of missing data from ? to NA
```{r}
genotypes_melt[genotypes_melt == "?/?"] = NA
```

Create a new column for homozygotes using a logical statement of TRUE, FALSE, NA for the missing data. Then sort the data by Sample_ID and Group.
```{r}
genotypes_melt$Homozygote <- (genotypes_melt$value == "A/A" | genotypes_melt$value == "C/C" | genotypes_melt$value == "T/T" | genotypes_melt$value == "G/G")

genotypes_melt <- arrange(genotypes_melt, Sample_ID, Group)
```

Plot # of homozygotes and heterozygotes by sample_ID
```{r}
counts_id <- ddply(genotypes_melt, c("Sample_ID"), summarise, Num_homozygous=sum(Homozygote, na.rm=TRUE), Num_heterozygous=sum(!Homozygote, na.rm=TRUE), Num_NA=sum(is.na(Homozygote)))

counts_id_melt <- melt(counts_id, measure.vars = c("Num_homozygous", "Num_heterozygous", "Num_NA"))
colnames(counts_id_melt)[3] <- "Counts"

ggplot(counts_id_melt, aes(x = Sample_ID, y= Counts, fill=variable)) + geom_bar(stat = "identity", position = "stack")

```

Plot homozygotes and heterozygotes by group
```{r}
counts_group <- ddply(genotypes_melt, c("Group"), summarise, Num_homozygous=sum(Homozygote, na.rm=TRUE), Num_heterozygous=sum(!Homozygote, na.rm=TRUE), Num_NA=sum(is.na(Homozygote)))

counts_group_melt <- melt(counts_group, measure.vars = c("Num_homozygous", "Num_heterozygous", "Num_NA"))

ggplot(counts_group_melt, aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack")
```

### Own Data Visualization - Most common genotypes across loci and individuals

```{r}
ggplot(genotypes_melt)+ geom_bar(aes(genotypes_melt$value))+xlab("Nucleotide") +ylab("Number of observations")
```



### Own Data Visualization- Recombination Rate per Sample_ID

Compute recombination rate as total number of heterozygous loci over number of homozygous and heterozygous loci.

```{r}
hetero_per_locus <- ddply(genotypes_melt, c("Sample_ID"), summarise, hetero_count=sum(!Homozygote, na.rm=TRUE), total_count=sum(!Homozygote, Homozygote, na.rm=TRUE))

hetero_per_locus$Recombination <- (hetero_per_locus$hetero_count/hetero_per_locus$total_count)

ggplot(hetero_per_locus, aes(x = Sample_ID, y= Recombination)) + geom_bar(stat = "identity", position = "stack")

```

